src/hierarchies/likelihoods/states/base_state.h

```diff
@@ -0,0 +1,35 @@
+#ifndef BAYESMIX_HIERARCHIES_LIKELIHOODS_STATES_BASE_STATE_H_
+#define BAYESMIX_HIERARCHIES_LIKELIHOODS_STATES_BASE_STATE_H_
+
+#include <memory>
+#include <stan/math/rev.hpp>
+
+#include "algorithm_state.pb.h"
+#include "src/utils/proto_utils.h"
+
+namespace State {
+
+class BaseState {
+ public:
+  int card;
+
+  using ProtoState = bayesmix::AlgorithmState::ClusterState;
+
+  virtual Eigen::VectorXd get_unconstrained() { return Eigen::VectorXd(0); }
+
+  virtual void set_from_unconstrained(Eigen::VectorXd in) {}
+
+  virtual void set_from_proto(const ProtoState &state_, bool update_card) = 0;
+
+  virtual ProtoState get_as_proto() = 0;
+
+  std::shared_ptr<ProtoState> to_proto() {
+    return std::make_shared<ProtoState>(get_as_proto());
+  }
+
+  virtual double log_det_jac() { return -1; }
```

Comment:
Any reason why some of these functions are not pure virtual?

Also you're missing const refs in `set_from_unconstrained()` and some const qualifiers for getters (and their overrides) 🤭

Reply:
> Any reason why some of these functions are not pure virtual?

I don't want to force users to implement the machinery to have the unconstrained representations since it's not always needed (e.g. if they use a custom SemiConjugateUpdater)